src/theory/strings/regexp_elim.cpp

```diff
@@ -0,0 +1,441 @@
+}
+
+Node RegExpElimination::eliminate(Node atom)
+{
+  Assert(atom.getKind() == STRING_IN_REGEXP);
+  if (atom[1].getKind() == REGEXP_CONCAT)
+  {
+    return eliminateConcat(atom);
+  }
+  else if (atom[1].getKind() == REGEXP_STAR)
+  {
+    return eliminateStar(atom);
+  }
+  return Node::null();
+}
+
+Node RegExpElimination::eliminateConcat(Node atom)
+{
+  NodeManager* nm = NodeManager::currentNM();
+  Node x = atom[0];
+  Node lenx = nm->mkNode(STRING_LENGTH, x);
+  Node re = atom[1];
+  // memberships of the form x in re.++ * s1 * ... * sn *, where * are
+  // any number of repetitions (exact or indefinite) of re.allchar.
+  Trace("re-elim-debug") << "Try re concat with gaps " << atom << std::endl;
+  std::vector<Node> children;
+  TheoryStringsRewriter::getConcat(re, children);
+  bool success = true;
+  std::vector<Node> sep_children;
+  std::vector<unsigned> gap_minsize;
+  std::vector<bool> gap_exact;
+  // the first gap is initially strict zero
+  gap_minsize.push_back(0);
+  gap_exact.push_back(true);
+  unsigned nchildren = children.size();
+  for (unsigned i = 0; i < nchildren; i++)
+  {
+    Node c = children[i];
+    Trace("re-elim-debug") << "  " << c << std::endl;
+    success = false;
+    if (c.getKind() == STRING_TO_REGEXP)
+    {
+      success = true;
+      sep_children.push_back(c[0]);
+      // the next gap is initially strict zero
+      gap_minsize.push_back(0);
+      gap_exact.push_back(true);
+    }
+    else if (c.getKind() == REGEXP_STAR && c[0].getKind() == REGEXP_SIGMA)
+    {
+      // found a gap of any size
+      success = true;
+      gap_exact[gap_exact.size() - 1] = false;
+    }
+    else if (c.getKind() == REGEXP_SIGMA)
+    {
+      // add one to the minimum size of the gap
+      success = true;
+      gap_minsize[gap_minsize.size() - 1]++;
+    }
+    if (!success)
+    {
+      Trace("re-elim-debug") << "...cannot handle " << c << std::endl;
+      break;
+    }
+  }
+  // we should always rewrite concatenations that are purely re.allchar
+  // and re.*( re.allchar ).
+  Assert(!success || !sep_children.empty());
+  if (success && !sep_children.empty())
+  {
+    std::vector<Node> conj;
+    // The following constructs a set of constraints that encodes that a
+    // set of string terms are found, in order, in string x.
+    // prev_end stores the current (symbolic) index in x that we are
+    // searching.
+    Node prev_end = d_zero;
+    unsigned gap_minsize_end = gap_minsize.back();
+    bool gap_exact_end = gap_exact.back();
+    std::vector<Node> non_greedy_find_vars;
+    for (unsigned i = 0, size = sep_children.size(); i < size; i++)
+    {
+      Node sc = sep_children[i];
+      if (gap_minsize[i] > 0)
+      {
+        // the gap to this child is at least gap_minsize[i]
+        prev_end =
+            nm->mkNode(PLUS, prev_end, nm->mkConst(Rational(gap_minsize[i])));
+      }
+      Node lensc = nm->mkNode(STRING_LENGTH, sc);
+      if (gap_exact[i])
```

Comment:
Shouldn't this condition be `gap_exact[i] && i + 1 != size`? Otherwise we add two `STRING_SUBSTR` constraints for the last index.

Reply:
I will refactor some of this, I believe there is a bug.